### PR TITLE
Changed the logic to be able to find the last IN event despite its order in the debug records.

### PR DESCRIPTION
**Pull request by irynamatveieva**
## Pull Request description

In case two most recent Debug events in a rule node contain two consecutive OUT messages with no IN, the getLatestRuleNodeDebugInput function can’t find the IN event to populate the test transformer function, so users get a default message and metadata fields:
image-20220902-075251
Changed the logic to be able to find the last IN event despite its order in the debug records. 

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/controller/RuleChainController.java [hunk 1]
```diff
@@ -340,18 +339,8 @@ public JsonNode getLatestRuleNodeDebugInput(
...
-                JsonNode body = event.getBody();
-                if (body.has("type") && body.get("type").asText().equals("IN")) {
-                    result = body;
-                    break;
-                }
-            }
-        }
-        return result;
+        EventInfo eventInfo = eventService.findLatestDebugRuleNodeInEvent(tenantId, ruleNodeId);
+        return eventInfo.getBody();
```
Comment (ShvaykaD) [inline application/src/main/java/org/thingsboard/server/controller/RuleChainController.java:343]: check for null, if eventInfo is null you will get NPE.

File: common/dao-api/src/main/java/org/thingsboard/server/dao/event/EventService.java [hunk 2]
```diff
@@ -35,6 +35,8 @@ public interface EventService {
 
     List<EventInfo> findLatestEvents(TenantId tenantId, EntityId entityId, EventType eventType, int limit);
 
+    EventInfo findLatestDebugRuleNodeInEvent(TenantId tenantId, EntityId entityId);
```
Comment (ShvaykaD) [inline common/dao-api/src/main/java/org/thingsboard/server/dao/event/EventService.java:38]: Let's add optional here.
```suggestion
    Optional<EventInfo> findLatestDebugRuleNodeInEvent(TenantId tenantId, EntityId entityId);
```

File: dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java [hunk 3]
```diff
@@ -117,4 +138,13 @@ private Event getStatsEvent(UUID eventId, UUID tenantId, UUID entityId) {
         event.errorsOccurred(0);
         return event.build();
     }
+
+    private RuleNodeDebugEventEntity getDebugEventEntity(UUID eventId, UUID tenantId, UUID entityId) {
```
Comment (ShvaykaD) [inline dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java:142]:
```suggestion
    private RuleNodeDebugEventEntity getDebugInEventEntity(UUID eventId, UUID tenantId, UUID entityId) {
```

File: dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java [hunk 4]
```diff
@@ -106,6 +111,22 @@ public void findEventsByEntityIdAndPageLink() throws Exception {
 
     }
 
+    @Test
+    public void findLatestDebugRuleNodeInEvent() {
+
+        UUID entityId = Uuids.timeBased();
+
+        RuleNodeDebugEventEntity event = getDebugEventEntity(Uuids.timeBased(), tenantId, entityId);
+        eventDao.saveAsync(event.toData());
```
Comment (ShvaykaD) [inline dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java:120]:
```suggestion
        eventDao.saveAsync(event.toData()).get(1, TimeUnit.MINUTES);
```

File: dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java [hunk 5]
```diff
@@ -106,6 +111,22 @@ public void findEventsByEntityIdAndPageLink() throws Exception {
...
+        RuleNodeDebugEventEntity event2 = getDebugEventEntity(Uuids.timeBased(), tenantId, entityId);
+        eventDao.saveAsync(event2.toData());
```
Comment (ShvaykaD) [inline dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java:122]:
```suggestion
        eventDao.saveAsync(event2.toData()).get(1, TimeUnit.MINUTES);

File: dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java (hunk 4 above)
Comment (ShvaykaD) [inline dao/src/test/java/org/thingsboard/server/dao/sql/event/JpaBaseEventDaoTest.java:120]: use 2 millis sleep in the same way as in other tests.

**Review by ShvaykaD — CHANGES_REQUESTED** [review]
See my comments added above.

**Review by ShvaykaD — APPROVED** [review]

File: application/src/main/java/org/thingsboard/server/controller/RuleChainController.java [hunk 6]
```diff
@@ -340,18 +339,8 @@ public JsonNode getLatestRuleNodeDebugInput(
...
-                JsonNode body = event.getBody();
-                if (body.has("type") && body.get("type").asText().equals("IN")) {
-                    result = body;
-                    break;
-                }
-            }
-        }
-        return result;
+        EventInfo eventInfo = eventService.findLatestDebugRuleNodeInEvent(tenantId, ruleNodeId);
+        return eventInfo == null ? null : eventInfo.getBody();
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/controller/RuleChainController.java:343]: Use `Optional.ofNullable`

File: dao/src/main/java/org/thingsboard/server/dao/event/BaseEventService.java [hunk 7]
```diff
@@ -140,4 +145,8 @@ private List<EventInfo> convert(EntityType entityType, List<? extends Event> lis
         return list == null ? null : list.stream().map(e -> e.toInfo(entityType)).collect(Collectors.toList());
     }
 
+    private EventInfo convert(EntityType entityType, Event event) {
+        return event == null ? null : event.toInfo(entityType);
```
Comment (ShvaykaD) [inline dao/src/main/java/org/thingsboard/server/dao/event/BaseEventService.java:149]: Please use Optional.ofNullable

**Pull request merged**